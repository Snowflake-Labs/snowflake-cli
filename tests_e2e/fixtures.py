import shutil
from contextlib import contextmanager
from pathlib import Path

import pytest


@pytest.fixture
def project_directory(temp_dir, test_root_path):
    @contextmanager
    def _temporary_project_directory(project_name):
        test_data_file = test_root_path / "test_data" / project_name
        shutil.copytree(test_data_file, temp_dir, dirs_exist_ok=True)
        yield Path(temp_dir)

    return _temporary_project_directory
