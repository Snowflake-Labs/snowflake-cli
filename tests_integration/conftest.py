from __future__ import annotations

import functools
import pytest
import toml

from pathlib import Path
from snowcli.cli import app
from tempfile import NamedTemporaryFile
from typer import Typer
from typer.testing import CliRunner

TEST_DIR = Path(__file__).parent


@pytest.fixture(scope="session")
def test_snowcli_config():
    test_config = TEST_DIR / "config/test.toml"
    config = toml.load(test_config)
    config["snowsql_config_path"] = str(TEST_DIR / "config/connection_configs.toml")
    with NamedTemporaryFile(suffix=".toml", mode="w+") as fh:
        toml.dump(config, fh)
        fh.flush()
        yield fh.name


class SnowCLIRunner(CliRunner):
    def __init__(self, app: Typer, test_snowcli_config: str):
        super().__init__()
        self.app = app
        self.test_snowcli_config = test_snowcli_config

    @functools.wraps(CliRunner.invoke)
    def invoke(self, *a, **kw):
        return super().invoke(self.app, *a, **kw)

    def invoke_with_config(self, *args, **kwargs):
        return self.invoke(
            ["--config-file", self.test_snowcli_config, *args[0]], **kwargs
        )


@pytest.fixture
def runner(test_snowcli_config):
    return SnowCLIRunner(app, test_snowcli_config)
