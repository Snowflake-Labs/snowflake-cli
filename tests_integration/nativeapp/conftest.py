from contextlib import contextmanager

import pytest

from tests_integration.conftest import SnowCLIRunner


@pytest.fixture
def nativeapp_project_directory(project_directory, nativeapp_teardown):
    @contextmanager
    def _nativeapp_project_directory(name):
        with project_directory(name) as d:
            with nativeapp_teardown():
                yield d

    return _nativeapp_project_directory


@pytest.fixture
def nativeapp_teardown(runner: SnowCLIRunner):
    @contextmanager
    def _nativeapp_teardown():
        try:
            yield
        finally:
            result = runner.invoke_with_connection_json(
                ["app", "teardown", "--force", "--cascade"]
            )
            assert result.exit_code == 0

    return _nativeapp_teardown
