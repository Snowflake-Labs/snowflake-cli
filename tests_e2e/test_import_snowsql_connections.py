import json
import os
from unittest import mock

import pytest

from tests_e2e.conftest import subprocess_check_output, subprocess_run


@pytest.fixture()
def _assert_json_output_matches_snapshot(snapshot):
    def f(cmd):
        output = subprocess_check_output(cmd)
        parsed_json = json.loads(output)
        snapshot.assert_match(json.dumps(parsed_json))

    return f


@pytest.mark.e2e
@mock.patch.dict(os.environ, {}, clear=True)
def test_import_of_snowsql_connections(
    snowcli, test_root_path, empty_config_file, _assert_json_output_matches_snapshot
):
    _assert_json_output_matches_snapshot(
        [
            snowcli,
            "--config-file",
            empty_config_file,
            "connection",
            "list",
            "--format",
            "json",
        ],
    )

    _assert_json_output_matches_snapshot(
        [
            snowcli,
            "--config-file",
            empty_config_file,
            "helpers",
            "import-snowsql-connections",
            "--snowsql-config-file",
            test_root_path / "config" / "snowsql" / "config",
            "--snowsql-config-file",
            test_root_path / "config" / "snowsql" / "overriding_config",
            "--format",
            "json",
        ],
    )

    _assert_json_output_matches_snapshot(
        [
            snowcli,
            "--config-file",
            empty_config_file,
            "connection",
            "list",
            "--format",
            "json",
        ]
    )


@pytest.mark.e2e
@mock.patch.dict(os.environ, {}, clear=True)
def test_import_error_on_conflict_with_existing_cli_connection(
    snowcli, test_root_path, config_file, snapshot
):
    result = subprocess_run(
        [
            snowcli,
            "--config-file",
            config_file,
            "helpers",
            "import-snowsql-connections",
            "--snowsql-config-file",
            test_root_path / "config" / "snowsql" / "integration_config",
            "--format",
            "json",
        ],
    )

    assert result.returncode == 1
    snapshot.assert_match(result.stderr)


@pytest.mark.e2e
def test_connection_imported_from_snowsql(
    snowcli, test_root_path, empty_config_file, _assert_json_output_matches_snapshot
):
    _assert_json_output_matches_snapshot(
        [
            snowcli,
            "--config-file",
            empty_config_file,
            "helpers",
            "import-snowsql-connections",
            "--snowsql-config-file",
            test_root_path / "config" / "snowsql" / "integration_config",
            "--format",
            "json",
        ],
    )

    result = subprocess_run(
        [
            snowcli,
            "--config-file",
            empty_config_file,
            "connection",
            "test",
            "-c",
            "integration",
            "--format",
            "json",
        ],
    )
    assert result.returncode == 0
