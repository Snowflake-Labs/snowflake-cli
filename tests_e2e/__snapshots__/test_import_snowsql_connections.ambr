# serializer version: 1
# name: test_connection_imported_from_snowsql
  '{"message": "Connections successfully imported from SnowSQL to Snowflake CLI."}'
# ---
# name: test_import_error_on_conflict_with_existing_cli_connection
  '''
  +- Error ----------------------------------------------------------------------+
  | Cannot import connections from SnowSQL because some of them conflict with    |
  | existing CLI's connections. Please remove or rename the following            |
  | connections: integration.                                                    |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_import_of_snowsql_connections
  '[]'
# ---
# name: test_import_of_snowsql_connections.1
  '{"message": "Connections successfully imported from SnowSQL to Snowflake CLI."}'
# ---
# name: test_import_of_snowsql_connections.2
  '[{"connection_name": "snowsql1", "parameters": {"account": "a1", "user": "u1", "host": "h1_override", "database": "d1", "schema": "public", "warehouse": "w1", "role": "r1"}, "is_default": false}, {"connection_name": "snowsql2", "parameters": {"account": "a2", "user": "u2", "host": "h2", "port": 1234, "database": "d2", "schema": "public", "warehouse": "w2", "role": "r2"}, "is_default": false}, {"connection_name": "example", "parameters": {"account": "accountname", "user": "username"}, "is_default": false}, {"connection_name": "snowsql3", "parameters": {"account": "a3", "user": "u3", "password": "****", "host": "h3", "database": "d3", "schema": "public", "warehouse": "w3", "role": "r3"}, "is_default": false}, {"connection_name": "default", "parameters": {"account": "default_connection_account", "user": "default_connection_user", "host": "localhost", "database": "default_connection_database_override", "schema": "public", "warehouse": "default_connection_warehouse", "role": "accountadmin"}, "is_default": true}]'
# ---
